util/EIP-712/src/encode.rs

```diff
@@ -0,0 +1,414 @@
+// Copyright 2015-2018 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! EIP712 Encoder
+use ethabi::{encode, Token as EthAbiToken};
+use ethereum_types::{Address as EthAddress, U256, H256};
+use keccak_hash::keccak;
+use serde_json::Value;
+use std::str::FromStr;
+use itertools::Itertools;
+use linked_hash_set::LinkedHashSet;
+use serde_json::to_value;
+use parser::{Parser, Type};
+use error::{Result, ErrorKind, serde_error};
+use eip712::{EIP712, MessageTypes};
+use rustc_hex::FromHex;
+use validator::Validate;
+use std::collections::HashSet;
+
+/// given a type and HashMap<String, Vec<FieldType>>
+/// returns a HashSet of dependent types of the given type
+fn build_dependencies<'a>(message_type: &'a str, message_types: &'a MessageTypes) -> Option<(HashSet<&'a str>)>
+{
+	if message_types.get(message_type).is_none() {
+		return None;
+	}
+
+	let mut types = LinkedHashSet::new();
```

Comment:
[indexmap](https://github.com/bluss/indexmap) should perform better here (and it does have `pop`), although it's definitely not a bottleneck.